controlplane/rosa/controllers/rosacontrolplane_controller.go

```diff
@@ -0,0 +1,387 @@
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	rosaScope, err := scope.NewROSAControlPlaneScope(scope.ROSAControlPlaneScopeParams{
+		Client:         r.Client,
+		Cluster:        cluster,
+		ControlPlane:   rosaControlPlane,
+		ControllerName: strings.ToLower(rosaControlPlaneKind),
+	})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to create scope: %w", err)
+	}
+
+	// Always close the scope
+	defer func() {
+		if err := rosaScope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	if !rosaControlPlane.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Handle deletion reconciliation loop.
+		return r.reconcileDelete(ctx, rosaScope)
+	}
+
+	// Handle normal reconciliation loop.
+	return r.reconcileNormal(ctx, rosaScope)
+}
+
+func (r *ROSAControlPlaneReconciler) reconcileNormal(ctx context.Context, rosaScope *scope.ROSAControlPlaneScope) (res ctrl.Result, reterr error) {
+	rosaScope.Info("Reconciling ROSAControlPlane")
+
+	// if !rosaScope.Cluster.Status.InfrastructureReady {
+	//	rosaScope.Info("Cluster infrastructure is not ready yet")
+	//	return ctrl.Result{RequeueAfter: r.WaitInfraPeriod}, nil
+	//}
+	if controllerutil.AddFinalizer(rosaScope.ControlPlane, ROSAControlPlaneFinalizer) {
+		if err := rosaScope.PatchObject(); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// TODO:
+	// - Implement deletion.
+	// - Revisit API input.
+	// - Suite Machine controller to account for ROSACluster resource.
+	// - Explore machine management, e.g capi MachinePool implementation to talk to OCM MachinePool API.
+
+	// Create the cluster:
+	clusterBuilder := cmv1.NewCluster().
+		Name(rosaScope.ControlPlane.Name).
+		MultiAZ(true).
+		Product(
+			cmv1.NewProduct().
+				ID("rosa"),
+		).
+		Region(
+			cmv1.NewCloudRegion().
+				ID(*rosaScope.ControlPlane.Spec.Region),
+		).
+		FIPS(false).
+		EtcdEncryption(false).
+		DisableUserWorkloadMonitoring(true).
+		Version(
+			cmv1.NewVersion().
+				ID(*rosaScope.ControlPlane.Spec.Version).
+				ChannelGroup("stable"),
+		).
+		ExpirationTimestamp(time.Now().Add(1 * time.Hour)).
+		Hypershift(cmv1.NewHypershift().Enabled(true))
+
+	networkBuilder := cmv1.NewNetwork()
+	networkBuilder = networkBuilder.Type("OVNKubernetes")
+	networkBuilder = networkBuilder.MachineCIDR(*rosaScope.ControlPlane.Spec.MachineCIDR)
+	clusterBuilder = clusterBuilder.Network(networkBuilder)
+
+	stsBuilder := cmv1.NewSTS().RoleARN(*rosaScope.ControlPlane.Spec.InstallerRoleARN)
+	// stsBuilder = stsBuilder.ExternalID(config.ExternalID)
+	stsBuilder = stsBuilder.SupportRoleARN(*rosaScope.ControlPlane.Spec.SupportRoleARN)
+	roles := []*cmv1.OperatorIAMRoleBuilder{}
+	for _, role := range []struct {
+		Name      string
+		Namespace string
+		RoleARN   string
+		Path      string
+	}{
+		{
+			Name:      "cloud-credentials",
+			Namespace: "openshift-ingress-operator",
+			RoleARN:   rosaScope.ControlPlane.Spec.RolesRef.IngressARN,
```

Comment:
right, they should exist already.
OCM SDK would report missing roles on cluster creation. Although it would be a nice enhancement to report that early on in a condition.